pkg/controller/cloudinit/controller.go

```diff
@@ -0,0 +1,231 @@
+package cloudinit
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+	"time"
+
+	ctlnodev1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	cloudinitv1 "github.com/harvester/node-manager/pkg/apis/node.harvesterhci.io/v1beta1"
+	"github.com/harvester/node-manager/pkg/cloudinit"
+	ctrlv1 "github.com/harvester/node-manager/pkg/generated/controllers/node.harvesterhci.io/v1beta1"
+)
+
+const handlerName = "harvester-node-cloud-init-controller"
+
+type controller struct {
+	nodeName   string
+	cloudinits ctrlv1.CloudInitClient
+	nodeCache  ctlnodev1.NodeCache
+}
+
+func Register(ctx context.Context, nodeName string, cloudinits ctrlv1.CloudInitController, nodeCache ctlnodev1.NodeCache) {
+	ctl := &controller{
+		nodeName:   nodeName,
+		cloudinits: cloudinits,
+		nodeCache:  nodeCache,
+	}
+
+	cloudinits.OnChange(ctx, handlerName, ctl.OnCloudInitChange)
+}
+
+func (c *controller) OnCloudInitChange(_ string, cloudInitObj *cloudinitv1.CloudInit) (*cloudinitv1.CloudInit, error) {
+	if cloudInitObj == nil || cloudInitObj.DeletionTimestamp != nil {
+		return cloudInitObj, nil
+	}
+
+	cloudInitCopy := cloudInitObj.DeepCopy()
+
+	if cloudInitCopy.Annotations == nil {
+		cloudInitCopy.Annotations = make(map[string]string)
+	}
+
+	checksum, err := cloudinit.Measure(strings.NewReader(cloudInitCopy.Spec.Contents))
+	_ = err
+
+	checksumString := fmt.Sprintf("%x", checksum)
+	persistedChecksum := strings.ToLower(cloudInitCopy.Annotations[cloudinit.AnnotationHash])
+
+	if checksumString != persistedChecksum {
+		cloudInitCopy.Annotations[cloudinit.AnnotationHash] = checksumString
+		return c.cloudinits.Update(cloudInitCopy)
+	}
+
+	node, err := c.nodeCache.Get(c.nodeName)
+	if err != nil {
+		return cloudInitCopy, err
+	}
+
+	if cloudinit.MatchesNode(node, cloudInitCopy) {
+		_ = cloudinit.RequireLocal(cloudInitCopy)
+	}
+
+	return c.updateStatus(node, cloudInitCopy)
```

Comment:
we should re-organise this a bit. If a node does not match the condition, we should return earlier.

currently when a node is not matching the filters, even then we are going through the process of evaluating all the conditions.

if notApplicable condition is set in the first method,  https://github.com/harvester/node-manager/pull/15/files#diff-8ad79d727a2364fb392ad1a30a059ed49322057d1bcc8788eefb0f83737ba154R107-R111, it would be safe to assume none of the other conditinos will ever be true. may be wise to apply those conditions and update status and return quickly?

Reply:
Agreed, will incorporate that.

Reply:
It occurred to me when I went to update this during the rebase that there is a situation in which we'd want to process all of the conditions, and ultimately I think it is simplest to do so unconditionally.

For example if the CloudInit resource's MatchSelector is updated and that results in one or more nodes being excluded, all 3 conditions can be used as valuable hints to the operator to know why a node is still being customized a certain way.

I.e., Applicable == False, but Present == True suggests the controller's automatic removal failed and the operator can get things back the way they expect it to by SSH'ing in and manually removing it, etc.